### PR TITLE
Add GHA CI & Docker 

### PR DESCRIPTION
**Pull request by rhaschke**
Adds GitHub Actions for creation of Docker images with pre-installed MuJoCo versions, and CI build and testing with CodeCov

## Todos
* Github Apps:
  - [ ] Install behavior bot for this repo (https://github.com/behaviorbot/welcome)
  - [ ] Install CodeCov for this repo (https://github.com/apps/codecov)
* Docker
  - [x] Change dockerhub user
  - [x] Add dockerhub secrets
* Formatting
 ~Apply formatting result patch to PR~
* Docs
  - [x] Change README.md badge links
* PR Rules
  - [ ] Require successful Format and CI Checks

### CODE REVIEW

File: .docker/mujoco_installer.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+rm -rf "${HOME}/mujoco"
+
+for mj_version in $MUJOCO_VERSIONS
+do
+	echo "Installing MuJoCo Version ${mj_version} ..."
+	curl -O -L https://github.com/google-deepmind/mujoco/releases/download/${mj_version}/mujoco-${mj_version}-linux-x86_64.tar.gz || echo "could not find version ${mj_version}!"
```
Comment (DavidPL1) [inline .docker/mujoco_installer.sh]: This should be preceded by 

```bash
available=$(curl -o /dev/null --silent -Iw '%{http_code}' \
		https://github.com/google-deepmind/mujoco/releases/download/${mj_version}/mujoco-${mj_version}-linux-x86_64.tar.gz)
	if [[ $available -eq 404 ]]; then
		echo "version ${mj_version} could not be found!"
		continue
	fi
```

Otherwise, non-existing version numbers are redirected to a 404 page stating "Not Found" which is still downloaded and will cause

File: .docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
+ARG ROS_DISTRO=noetic
+FROM ros:${ROS_DISTRO}-ros-base
+
+ARG MUJOCO_VERSIONS=[2.3.6]
```
Comment (DavidPL1) [inline .docker/Dockerfile]: Because of the changed script the brackets should be removed

File: .docker/mujoco_installer.sh (hunk 1 above)
Comment (rhaschke) [reply .docker/mujoco_installer.sh]: Better use option -f (--fail).

File: mujoco_ros/CMakeLists.txt [hunk 3]
```diff
@@ -45,7 +45,7 @@ find_package(catkin REQUIRED
     image_transport
     camera_info_manager
     sensor_msgs
-    moveit_ros_planning_interface
+    # moveit_ros_planning_interface
```
Comment (DavidPL1) [inline mujoco_ros/CMakeLists.txt]: moveit_ros_planning_interface can be fully removed from CMakeLists.txt and package.xml

File: mujoco_ros/CMakeLists.txt
Comment (DavidPL1) [inline mujoco_ros/CMakeLists.txt:1]: For more catkin lint fixes see today's changes in [this](https://github.com/ubi-agni/mujoco_ros_pkgs/compare/noetic-devel...fix-catkin_lint) commit.

File: mujoco_ros_sensors/CMakeLists.txt [hunk 4]
```diff
@@ -4,16 +4,13 @@ project(mujoco_ros_sensors)
...
-# Disable optimization to prevent false-positives with memory analisys tools (e.g. asan or valgrind)
-if (ENABLE_SANITIZER OR NO_OPTIM)
-  set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} -O0 -fno-inline")
-  set(CMAKE_BUILD_TYPE "RelWithDebInfo")
+if(NOT CMAKE_BUILD_TYPE)
+  message(STATUS "CMAKE_BUILD_TYPE not set. Defaulting to 'Release'.")
+  set(CMAKE_BUILD_TYPE "Release")
 endif()
 
 if (ENABLE_SANITIZER)
```
Comment (DavidPL1) [inline mujoco_ros_sensors/CMakeLists.txt]: The mentioned commit also introduces the Sanitizer build type for the sensors and the control packages' CMakeLists

**Comment by rhaschke** [comment]
I had a hard time finding a compiler error only occurring with clang and only on GHA, but not locally (first).
Eventually, clang uses C++14 by default, while gcc already uses C++17, which is required by ROS/log4cxx.

Now, there are only real issues in the source code, revealed by the more picky clang compiler...
@DavidPL1, I hand over to you. 

I suggest to squash-merge this PR if ready.

File: mujoco_ros/include/mujoco_ros/plugin_utils.h [hunk 5]
```diff
@@ -51,7 +51,7 @@ class MujocoPlugin
 	void init(const XmlRpc::XmlRpcValue &config, ros::NodeHandlePtr nh, MujocoEnvPtr env_ptr)
```
Comment (rhaschke) [inline mujoco_ros/include/mujoco_ros/plugin_utils.h]: ros::NodeHandle doesn't need to be a pointer. It is a slim datastructure anyway.

**Review by rhaschke — COMMENTED** [review]
Any idea, why some unit tests got broken?

**Comment by DavidPL1** [comment]
> Any idea, why some unit tests got broken?

I'm not really sure yet. The last commit at least fixes _mujoco_ros_plugin_test_ on focal (with a catkin workspace configured to use the installspace) when run locally with 
```
roslaunch mujoco_ros_pkgs/mujoco_ros/test/launch/mujoco_ros_plugin.test 
devel/.private/mujoco_ros/lib/mujoco_ros/mujoco_ros_plugin_test
``` 
in two separate terminals.

However, running catkin test mujoco_ros still fails without any helpful information: 
```
[Testcase: testmujoco_ros_plugin_test] ... FAILURE!
FAILURE: False is not true : test [mujoco_ros_plugin_test] did not generate test results
```
And running colcon test gives the same output.

File: mujoco_ros/src/plugin_utils.cpp [hunk 6]
```diff
@@ -97,16 +97,17 @@ bool registerPlugin(ros::NodeHandlePtr nh, XmlRpc::XmlRpcValue &config, std::vec
 
 	ROS_DEBUG_STREAM_NAMED("mujoco_ros_plugin_loader", "Registering plugin of type " << type);
 
+	std::string ns(nh->getNamespace());
```
Comment (rhaschke) [inline mujoco_ros/src/plugin_utils.cpp]:
```suggestion
	const std::string& ns = nh->getNamespace();
```

File: mujoco_ros/test/CMakeLists.txt [hunk 7]
```diff
@@ -45,9 +45,13 @@ add_dependencies(mujoco_ros_plugin_test
...
 
 target_link_libraries(mujoco_ros_plugin_test
-  GTest::gtest_main
   mujoco_ros
   mujoco_testplugin
   project_option
   project_warning
 )
+
+install(FILES
```
Comment (rhaschke) [inline mujoco_ros/test/CMakeLists.txt:54]: The test worked without installation before.

File: mujoco_ros/test/test_plugin/CMakeLists.txt [hunk 8]
```diff
@@ -15,5 +15,5 @@ install(TARGETS mujoco_testplugin
 )
 install(FILES
   test_plugin.xml
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/test/test_plugin
```
Comment (rhaschke) [inline mujoco_ros/test/test_plugin/CMakeLists.txt:18]: The test worked without installation before. Why did you change your mind w.r.t. installation of the test plugin?

Comment (DavidPL1) [reply mujoco_ros/test/test_plugin/CMakeLists.txt:18]: Turns out it was being installed already, but into package's share root, which mismatched the definition in the package.xml export path. This only worked in the devel space, because the src path to the plugin xml matched the package.xml export.

The plugin test relies on mujoco_ros to load the plugin, thus it needs to a) find the plugin xml and b) load the plugin library defined in the plugin xml.
I still don't like the fact that these artifacts are being installed into the installspace, but at least this should only happen when building tests is enabled.

**Comment by DavidPL1** [comment]
Looks like were almost done.

The remaining clang-tidy errors now come from lodepng which currently is just a 1:1 copy of the original upstream header and implementation. 

There are two ways to handle this: 
Either we stick to the current version, apply the fixes and track upstream changes into our version (there haven't been many changes in the past year ([commit-history](https://github.com/lvandeve/lodepng/commits/master)))
or we switch to using CMake's FetchContent like deepmind does to fetch and build a specific commit upon building, and disable clang-tidy checks for the fetched contents.

What do you think, @rhaschke?

File: mujoco_ros/include/mujoco_ros/plugin_utils.h (hunk 5 above)
Comment (DavidPL1) [reply mujoco_ros/include/mujoco_ros/plugin_utils.h]: Keeping in mind that we might have a lot of plugins loaded in certain scenarios, does it make sense to switch to ros::NodeHandle?

**Comment by rhaschke** [comment]
> Either we stick to the current version, apply the fixes, and track upstream changes into our version ...

Is there a special reason to use lodepng instead of libpng? According to [this blog](https://blog.gibson.sh/2015/03/23/comparing-performance-stb_image-vs-libjpeg-turbo-libpng-and-lodepng/), the latter is much faster.
In general, I don't care about the way to go. I think adapting the imported files is the easiest.

File: mujoco_ros/include/mujoco_ros/plugin_utils.h (hunk 5 above)
Comment (rhaschke) [reply mujoco_ros/include/mujoco_ros/plugin_utils.h]: I guess you are asking whether or not to change public API?
I think, it's still time to do so ;-) If you wait longer, it become even more difficult.

Comment (DavidPL1) [reply mujoco_ros/include/mujoco_ros/plugin_utils.h]: The API is changed anyway because of several switches from shared pointers to raw pointers in the callbacks, so now would be the perfect time for such a change.

My question is if there is a benefit in copying the NodeHandle for each plugin rather than using shared pointers.

**Comment by DavidPL1** [comment]
> > Either we stick to the current version, apply the fixes, and track upstream changes into our version ...
> 
> Is there a special reason to use lodepng instead of libpng? According to [this blog](https://blog.gibson.sh/2015/03/23/comparing-performance-stb_image-vs-libjpeg-turbo-libpng-and-lodepng/), the latter is much faster. In general, I don't care about the way to go. I think adapting the imported files is the easiest.

There is no specific reason other than using the same library as deepmind does in their simulator program. They also don't explain their choice.
The test you linked is rather old (2015) and the more recent comments don't mention newer results for lodepng. Additionally, the benchmark only covers image decoding. 

In this case lodepng is only used to write a PNG from a raw RGB buffer when clicking the screenshot button, which is also not expected to happen too often. Keeping the library thin seems to be prioritized over performance here.

File: mujoco_ros/include/mujoco_ros/lodepng.h [hunk 9]
```diff
@@ -118,8 +122,7 @@ or comment out LODEPNG_COMPILE_CPP below*/
 
 #ifdef LODEPNG_COMPILE_PNG
 /*The PNG color types (also used for raw image).*/
-typedef enum LodePNGColorType
-{
+using LodePNGColorType = enum LodePNGColorType {
```
Comment (rhaschke) [inline mujoco_ros/include/mujoco_ros/lodepng.h:125]: I think you can omit the using statement in all these type definitions as the type is already directly accessible with the given name.

**Review by rhaschke — COMMENTED** [review]
Looks good. I think you can squash-merge :1st_place_medal:

File: mujoco_ros/include/mujoco_ros/lodepng.h (hunk 9 above)
Comment (DavidPL1) [reply mujoco_ros/include/mujoco_ros/lodepng.h:125]: I tried removing the using statements but that causes C errors during compilation. I'll leave it as is for now

**Comment by DavidPL1** [comment]
> Looks good. I think you can squash-merge 🥇

I'll prepare the other plugins outside of this repo to comply with the API changes and then squash-merge the PR. Thanks for your help, especially with the workflows!

**Comment by rhaschke** [comment]
You are welcome.
